daemon/daemon_test.go

```diff
@@ -314,6 +315,10 @@ func (s *daemonSuite) TestMaintenanceJsonDeletedOnStart(c *check.C) {
 		Message: systemRestartMsg,
 	}
 
+	// mock testing environment to avoid the shutdown wait
+	r := snapdenv.MockTesting(true)
```

Comment:
I wonder if a slightly different approach here might be simpler? Just mocking the global rebootNoticeWait, something like?
```go
diff --git a/daemon/daemon_test.go b/daemon/daemon_test.go
index fa431bac98..a3c73420e9 100644
--- a/daemon/daemon_test.go
+++ b/daemon/daemon_test.go
@@ -51,7 +51,6 @@ import (
 	"github.com/snapcore/snapd/overlord/state"
 	"github.com/snapcore/snapd/snap"
 	"github.com/snapcore/snapd/snap/snaptest"
-	"github.com/snapcore/snapd/snapdenv"
 	"github.com/snapcore/snapd/store"
 	"github.com/snapcore/snapd/systemd"
 	"github.com/snapcore/snapd/testutil"
@@ -84,6 +83,7 @@ func (s *daemonSuite) SetUpTest(c *check.C) {
 	}
 	s.notified = nil
 	s.AddCleanup(ifacestate.MockSecurityBackends(nil))
+	s.AddCleanup(MockRebootNoticeWait(0))
 }
 
 func (s *daemonSuite) TearDownTest(c *check.C) {
@@ -315,10 +315,6 @@ func (s *daemonSuite) TestMaintenanceJsonDeletedOnStart(c *check.C) {
 		Message: systemRestartMsg,
 	}
 
-	// mock testing environment to avoid the shutdown wait
-	r := snapdenv.MockTesting(true)
-	defer r()
-
 	b, err := json.Marshal(maintErr)
 	c.Assert(err, check.IsNil)
 	c.Assert(os.MkdirAll(filepath.Dir(dirs.SnapdMaintenanceFile), 0755), check.IsNil)
@@ -670,10 +666,6 @@ version: 1`, si)
 func (s *daemonSuite) TestRestartWiring(c *check.C) {
 	d := s.newTestDaemon(c)
 
-	// mock testing environment to avoid the shutdown wait
-	r := snapdenv.MockTesting(true)
-	defer r()
-
 	// mark as already seeded
 	s.markSeeded(d)
 
@@ -749,10 +741,6 @@ func (s *daemonSuite) TestGracefulStop(c *check.C) {
 		}
 	})
 
-	// mock testing environment to avoid the shutdown wait
-	r := snapdenv.MockTesting(true)
-	defer r()
-
 	// mark as already seeded
 	s.markSeeded(d)
 	// and pretend we have snaps
@@ -839,10 +827,6 @@ version: 1`, si)
 func (s *daemonSuite) TestGracefulStopHasLimits(c *check.C) {
 	d := s.newTestDaemon(c)
 
-	// mock testing environment to avoid the shutdown wait
-	r := snapdenv.MockTesting(true)
-	defer r()
-
 	// mark as already seeded
 	s.markSeeded(d)
 
@@ -1357,10 +1341,6 @@ func (s *daemonSuite) TestRestartIntoSocketModeNoNewChanges(c *check.C) {
 	restore := standby.MockStandbyWait(5 * time.Millisecond)
 	defer restore()
 
-	// mock testing environment to avoid the shutdown wait
-	r := snapdenv.MockTesting(true)
-	defer r()
-
 	d := s.newTestDaemon(c)
 	makeDaemonListeners(c, d)
 
@@ -1390,10 +1370,6 @@ func (s *daemonSuite) TestRestartIntoSocketModePendingChanges(c *check.C) {
 	restore := standby.MockStandbyWait(5 * time.Millisecond)
 	defer restore()
 
-	// mock testing environment to avoid the shutdown wait
-	r := snapdenv.MockTesting(true)
-	defer r()
-
 	d := s.newTestDaemon(c)
 	makeDaemonListeners(c, d)
 
diff --git a/daemon/export_test.go b/daemon/export_test.go
index 27750df8f1..96dd32b6b2 100644
--- a/daemon/export_test.go
+++ b/daemon/export_test.go
@@ -329,3 +329,9 @@ func MockAspectstateSet(f func(databag aspects.DataBag, account, bundleName, asp
 		aspectstateSetAspect = old
 	}
 }
+
+func MockRebootNoticeWait(d time.Duration) (restore func()) {
+	restore = testutil.Backup(&rebootNoticeWait)
+	rebootNoticeWait = d
+	return restore
+}
```

Reply:
The main reason I am suggesting this is that with that we can check we can actually validate that the wait happend inside the unit test:
```go
@@ -1033,6 +1017,10 @@ func (s *daemonSuite) testRestartSystemWiring(c *check.C, prep func(d *Daemon),
 
 	err = d.Stop(nil)
 
+	// ensure Stop waited for at least rebootWaitTimeout
+	timeToStop := time.Now().Sub(now)
+	c.Check(timeToStop > rebootWaitTimeout+rebootNoticeWait, check.Equals, true)
+
 	c.Check(err, check.ErrorMatches, fmt.Sprintf("expected %s did not happen", expectedAction))
 
 	c.Check(delays, check.HasLen, 2)
`` `
attached is the full diff
[daemon_test-tweaks.diff.gz](https://github.com/snapcore/snapd/files/12038429/daemon_test-tweaks.diff.gz)

Reply:
Thank you for that suggestion, I think it's indeed more elegant, added this